source/auth/auth.rst

```diff
@@ -160,7 +160,8 @@ handshake:
    an authentication error.
 
 #. If the server is not of type Standalone, RSPrimary, RSSecondary or
-   Mongos, no authentication is possible and the handshake is complete.
+   Mongos and the connection is not a direct connection,
```

Comment:
Assuming we're going to add the arbiter exception:

Could we simply say that we authenticate unless it's an Arbiter?   For non-direct connections, is there any reason the driver would authenticate to any other type of server than the ones already listed?

Reply:
IIRC, single threaded drivers use one socket for monitoring and operations and therefore will normally auth to anything they connect to.  Other than direct connection (which is user's responsibility), we don't want them to attempt auth to RSOther/RSGhost.

Reply:
My practical question is: are multithreaded drivers like Java and C#, which just authenticate all pooled connections regardless of whether they are RSOther or RSSecondary (and whose handshakers are unaware of whether the connection is "direct" or not), in violation of this?  I would like the answer to be "no", and that the wording of the change makes that clear.

Reply:
I'm not sure I follow.  You said, "which just authenticate all pooled connections regardless" -- by that do you mean it just happens they never actually auth in practice because they aren't selected?  What would minPoolSize do?  What does it mean to have a minPoolSize to a non-data-bearing node (outside direct connections)?

Reply:
SDAM requires that, when is discovery mode, replica set primaries, secondaries, passives, and arbiters are included in the topology. But the built-in server selector (either for a writable server or to satisfy read preference) would only ever select a primary or secondary.  So yes, in normal practice passives and arbiters would never be selected and always have empty connection pools.

But the CMAP spec says nothing about treating passives and arbiters any differently.  In [this design rationale](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#why-do-we-set-minpoolsize-across-all-members-of-a-replicaset-when-most-traffic-will-be-against-a-primary), in fact, it specifically says not to treat different replica set member types any differently.  So you're absolutely right that if you set minPoolSize, drivers would end up creating connections to passives and arbiters in addition to primary and secondaries.  We could argue about whether that's the right behavior, but that's what it is now.

Given that specification, does it actually make sense to authenticate all server types (except arbiters, where auth would definitely fail) even when in discovery mode?  As an example, consider a replica set with a passive server that is, for some reason, not yet marked as a secondary; and an application with minPoolSize set:

* Driver creates minPoolSize connections to all servers, even the passive server
* Some time later the passive server becomes a secondary (perhaps after an initial sync completes?)
* Driver detects the change and updates the server type from RSOther to RSSecondary
* Server selection selects the new secondary for some read operation

Expected result: the pooled connections to the new secondary should be authenticated

Reply:
As clarified in https://github.com/mongodb/specifications/pull/645, pools should only be created for servers the drivers knows to be data bearing, i.e., those it may legitimately connect to to perform an operation.

https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#data-bearing-server-type includes standalone in the data bearing server list.

Reply:
Auth against a secondary doing initial sync might fail if the auth credentials haven't replicated.

Reply:
So be it.  Then the pool will be empty until authentication starts succeeding.  I think that's a reasonable behavior.